# 1. Định nghĩa về đệ quy

Đệ quy gồm ba mệnh đề:
+ **Mệnh đề cơ sở:** là mệnh đề không chứa khái niệm định nghĩa
+ **Mệnh đề đệ quy:** là mệnh đề có sử dụng lại khái niệm được định nghĩa
+ **Mệnh đề bao đóng:** là mệnh đề loại trừ những gì không liên quan đến mệnh đề cơ sở và mệnh đề đệ quy

Lưu ý: Một quá trình đệ quy là đúng đắn nếu quá trình này kết thúc bằng mệnh đề cơ sở nếu không thì chương trình sẽ lặp vô hạn.

Example 1: Tập E bao gồm các số chẵn không âm .

+ Số 0 thuộc tập E (mệnh đề cơ sở)

+ Nếu n thuộc tập E thì n+2 cũng thuộc tập E (mệnh đề đệ quy)

+ Không có số nào thuộc tập E, ngoại trừ tập E có được từ 2 mệnh đề trên (mệnh đề bao đóng)

Example 2: Tính giai thừa của n

+ 0! = 1 đây là mệnh đề cơ sở

+ n! = nx(n-1)! đây là mệnh đề đệ quy

+ n phải là một số nguyên không âm n>=0 đây là mệnh đề bao đóng

# 2. Hàm đệ quy
Một hàm đệ quy có một hoặc nhiều mệnh đề cơ sở và một hoặc nhiều mệnh đề đệ quy.

Sau mỗi lần đệ quy các dữ liệu vào được đơn giản hơn và lần gọi cuối cùng đạt tới mệnh đề cơ sở.

## Các loại đệ quy:

### + Đệ quy đơn: chỉ chứa một tự tham khảo (self- reference)

### + Đa đệ quy: chứa nhiều tự tham khảo
### + Đệ quy đuôi : là trường hợp đặc biệt của đệ quy mà việc gọi đệ quy là hành động cuối cùng
Example: 
+ Hàm tính giai thừa là đệ quy đơn vì thân hàm chỉ có một lần gọi đệ quy
+ Hàm tìm ucln là đệ quy đơn và đệ quy đuôi bởi vì gọi đệ quy là hành động cuối cùng.
+ Hàm Fibonaci là hàm đa đệ quy vì trong hàm có 2 lần gọi đệ quy.

In [1]:
#Đệ quy đơn
results=1
def factorial(n):
    if n==0:
        results=1
    else:
        results=n*factorial(n-1)
    return results
factorial(5)

120

Các thức đệ quy:

factorial(5)&nbsp;=&nbsp;5 \* factorial(4)
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 5 \* 4 \* factorial(3)
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 5 \* 4 \* 3 \* factorial(2)
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 5 \* 4 \* 3 \* 2 \* factorial(1)
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 5 \* 4 \* 3 \* 2 \* 1 \* factorial(0)
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 5 \* 4 \* 3 \* 2 \* 1 \* 1
             <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 120 

In [2]:
# Đệ quy đuôi
def gcd(a,b):
    if b==0:
        return a
    else:
        return gcd(b,a%b) 
gcd(2,19)

1

In [3]:
# Đa đệ quy
def Fibonaci(n):
    if n<=0:
        return 0
    elif n==1:
        return 1
    else:
        return Fibonaci(n-1)+Fibonaci(n-2)
Fibonaci(5)

5

**Sơ đồ cây của lời gọi hàm Fibonaci(5)**

<img align="left" width="500" height="500" src="images/Tree_Fibonaci.jpg">

Lưu ý: Trong thân của hàm đệ quy ta thường viết xử lý các mệnh đề cơ sở, sau đó ta viết đoạn mã xử lý các mệnh đề đệ quy. Mệnh đề bao đóng phải được kiểm tra trước khi gọi hàm đệ quy. Đệ quy thưởng kém hiệu quả do tốn bộ nhớ và thời gian chạy khá lâu sau mỗi lần truy hồi lại hàm.

### + Đệ quy gián tiếp: khi một hàm không được gọi chính nó mà bởi một hàm khác do nó gọi hàm này

Example: hàm F1 gọi hàm F2, hàm F2 gọi hàm F3 và hàm F3 gọi hàm F1. Hàm F1 gọi gián tiếp hàm F3.

In [4]:
def Even(n):
    if n==0:
        return True
    else:
        return Odd(n-1)
def Odd(n):
    if n==0:
        return False
    else:
        return Even(n-1)
Even(5)

False

# 3. Đệ quy và lặp vòng (Interation)

Đệ quy có thể thay thế bởi lặp vòng với cái chồng stack tường minh.

Lặp vòng có thể thay thế bởi đệ quy đuôi và ngược lại.

Người lập trình thường sử dụng giải thuật lặp vòng hoặc đệ quy đơn để tránh tốn nhiều thời gian gọi hàm.

Trong ngôn ngữ lập trình hàm người lập trình cũng thường sử dụng nhiều nhất là đệ quy đuôi.

Example 1: Tính giai thừa sử dụng đệ quy và lặp vòng.

In [5]:
#Đệ quy
results=1
def factorial(n):
    if n==0:
        results=1
    else:
        results=n*factorial(n-1)
    return results
factorial(5)

120

In [6]:
#Lặp vòng
def Factorial(n):
    factorial=1
    for i in range(2,n+1):
        factorial=factorial*i
    return factorial
Factorial(5)

120

Example 2: Tính ước chung lớn nhất của hai số sử dụng đệ quy và lặp vòng.

In [7]:
# Đệ quy
def gcd(a,b):
    if b==0:
        return a
    else:
        return gcd(b,a%b) 
gcd(8,18)

2

In [8]:
# Lặp vòng
def GCD(a,b):
    while b!=0:
        r=a%b
        a=b
        b=r
    return a
GCD(8,18)

2

In [9]:
# Sử dụng thuật toán naive
def iterative_gcd(num1, num2):
# Xác định giá trị lớn nhất của num1 và nùm
    Min = num1 if num1 < num2 else num2
# 1 được xem là factor ban đầu
    largest_factor = 1;
    for i in range(1, Min + 1):
        if num1 % i == 0 and num2 % i == 0:
            largest_factor = i
    return largest_factor
iterative_gcd(8,18)

2

Example 3: Tính số n của dãy Fibonaci sử dụng đệ quy và lặp vòng.

In [10]:
# Đệ quy
def Fibonaci(n):
    if n<=0:
        return 0
    elif n==1:
        return 1
    else:
        return Fibonaci(n-1)+Fibonaci(n-2)
Fibonaci(5)

5

In [11]:
# Lặp vòng
results=0
def Fibonaci(n):
    if n<=0:
        results=0
    elif n==1:
        results=1
    else:
        F2=0;F1=1
        for i in range(2, n+1):
            Fn=F2+F1
            F2=F1
            F1=Fn
        results=Fn
    return results
Fibonaci(5)

5

# 4. Quay lui (Backtracking)
Là một giải thuật xác định một chuỗi các điểm quyết định(decision point) mà mỗi chọn lựa sẽ dẫn đến một đường đi nào đó.
+ Nếu đường đi này cho ra kết quả mong muốn thì vấn đề đã được giải quyết.
+ Ngược lại thì quay lui về điểm quyết định trước đó để tìm đường khác.
+ Lặp lại cách này cho đến khi vấn đề được giải quyết.

<img align="left" width="500" height="500" src="images/Backtracking.jpg">

Example: Sinh hoán vị mảng a có n phần tử.

In [12]:
def permute(list_A, s):
    results=[]
    if list_A == 1:
        return s
    else:
        for y in permute(1,s):
            for x in permute(list_A-1,s):
                results.append(y+x)
    return results
print(permute(1, ["a","b","c"]))
print(permute(2, ["a","b","c"]))

['a', 'b', 'c']
['aa', 'ab', 'ac', 'ba', 'bb', 'bc', 'ca', 'cb', 'cc']


# 5. Chia để trị (divide and conquer)

Một bài toán lớn có thể được chia thành các bài toán con nhỏ hơn và sau đó mỗi bài toán được giải quyết một cách độc lập. 

Khi chúng ta tiếp tục chia các bài toán con thành các bài toán con thậm chí nhỏ hơn, cuối cùng chúng ta có thể đạt đến một giai đoạn mà không thể chia thêm nữa. 

Lời giải của tất cả các bài toán con cuối cùng được hợp nhất để có được lời giải của một bài toán ban đầu.

<img align="left" width="400" height="400" src="images/Divide_and_conquer.jpg">

Gồm 3 bước:

+ Divide
+ Conquer
+ Merge

Example: Sử dụng divide and conquer vào thuật toán tìm kiếm nhị phân với điều kiện mảng ban đầu đã được xắp xếp theo thứ tự tăng dần .

In [13]:
def binarysearch(list_A, key):
    id_end = len(list_A) - 1
    id_start = 0
    while id_start <= id_end:
        id_mid = (id_start + id_end)// 2
        if list_A[id_mid] == key:
            return id_mid
# So sánh key với giá trị giữa
        if key > list_A[id_mid]:
            id_start = id_mid + 1
        else:
            id_end = id_mid - 1
        if id_start > id_end:
            return None

list_A = [2,7,19,34,53,72]
print(binarysearch(list_A,72))
print(binarysearch(list_A,11))

5
None


Độ phức tạp của thuật toán này sẽ là O($log_2{n}$).

**Sử dụng đệ quy**

In [14]:
def bsearch(list_A, id_start, id_end, key):
    if (id_end < id_start):
        return None
    else:
        id_mid = id_start + ((id_end - id_start) // 2)
# Compare the search item with middle most value
    if list_A[id_mid] > key:
        return bsearch(list_A, id_start, id_mid-1,key)
    elif list_A[id_mid] < key:
        return bsearch(list_A, id_mid+1, id_end, key)
    else:
        return id_mid
list = [8,11,24,56,88,131]
print(bsearch(list, 0, 5, 24))
print(bsearch(list, 0, 5, 51))

2
None
